ultralist/input_parser.go

```diff
@@ -0,0 +1,195 @@
+
+# due today OR tomorrow
+due:tod,tom
+due:today due:tom
+
+# filter todos before a certain date
+due:<aug15
+
+completed:true
+
+priority:false
+
+project:one,-two
+
+*/
+
+// Parse parses raw input and returns a Filter object
+func (p *InputParser) Parse(input string) (*Filter, error) {
+	filter := &Filter{
+		HasStatus:        false,
+		HasCompleted:     false,
+		HasCompletedAt:   false,
+		HasIsPriority:    false,
+		HasProjectFilter: false,
+		HasContextFilter: false,
+		HasDueBefore:     false,
+		HasDue:           false,
+		HasDueAfter:      false,
+	}
+
+	dateParser := &DateParser{}
+
+	var subjectMatches []string
+
+	cr, _ := regexp.Compile(`\@[\p{L}\d_-]+`)
+	filter.Contexts = p.matchWords(input, cr)
+
+	pr, _ := regexp.Compile(`\+[\p{L}\d_-]+`)
+	filter.Projects = p.matchWords(input, pr)
+
+	for _, word := range strings.Split(input, " ") {
+		match := false
+		r1, _ := regexp.Compile(`archived:.*$`)
+		if r1.MatchString(word) {
+			filter.HasArchived = true
+			filter.Archived = p.parseBool(r1.FindString(word)[9:])
+			match = true
+		}
+
+		r2, _ := regexp.Compile(`priority:.*$`)
+		if r2.MatchString(word) {
+			filter.HasIsPriority = true
+			filter.IsPriority = p.parseBool(r2.FindString(word)[9:])
+			match = true
+		}
+
+		r3, _ := regexp.Compile(`completed:.*$`)
+		if r3.MatchString(word) {
+			filter.HasCompleted = true
+			filter.Completed = p.parseBool(r3.FindString(word)[10:])
+			match = true
+		}
+
+		rDueBefore, _ := regexp.Compile(`duebefore:.*$`)
+		if rDueBefore.MatchString(word) {
+			filter.HasDueBefore = true
+			dueDate, err := dateParser.ParseDate(rDueBefore.FindString(word)[10:], time.Now())
+			if err != nil {
+				return filter, err
+			}
+
+			if dueDate.IsZero() {
+				filter.DueBefore = ""
+			} else {
+				filter.DueBefore = dueDate.Format("2006-01-02")
+			}
+			match = true
+		}
+
+		rDue, _ := regexp.Compile(`due:.*$`)
+		if rDue.MatchString(word) {
+			filter.HasDue = true
+			dueDate, err := dateParser.ParseDate(rDue.FindString(word)[4:], time.Now())
+			if err != nil {
+				return filter, err
+			}
+
+			if dueDate.IsZero() {
+				filter.Due = ""
+			} else {
+				filter.Due = dueDate.Format("2006-01-02")
+			}
+			match = true
+		}
+
+		rDueAfter, _ := regexp.Compile(`dueafter:.*$`)
+		if rDueAfter.MatchString(word) {
+			filter.HasDueAfter = true
+			dueDate, err := dateParser.ParseDate(rDueAfter.FindString(word)[9:], time.Now())
+			if err != nil {
+				return filter, err
+			}
+
+			if dueDate.IsZero() {
+				filter.DueAfter = ""
+			} else {
+				filter.DueAfter = dueDate.Format("2006-01-02")
+			}
+			match = true
+		}
+
+		r5, _ := regexp.Compile(`status:.*$`)
+		if r5.MatchString(word) {
+			filter.HasStatus = true
+			filter.Status, filter.ExcludeStatus = p.parseString(r5.FindString(word)[7:])
+			match = true
+		}
+
+		r6, _ := regexp.Compile(`completedat:.*$`)
+		if r6.MatchString(word) {
+			filter.HasCompletedAt = true
+			filter.CompletedAt, _ = p.parseString(r6.FindString(word)[7:])
```

Comment:
can you check this, my gut is saying something is wrong here?

Reply:
yep, it was wrong.  also, I refactored this to re-use a single regex variable, rather than re-declaring `r1`,`r2`, etc.